test_python/test_tc.py

```diff
@@ -12,89 +12,499 @@
+        print("\nTune group_norm from naive options, don't store best options")
+        group_norm.tune(
+            I,
+            gamma,
+            beta,
+            mean.view((N, G)),
+            var.view((N, G)),
+            starting_options=tc.MappingOptions('naive'),
+            tuner_config=tuner_config)
+        print("\nRun tuned group_norm with best options")
+        out, = group_norm(I, gamma, beta, mean.view((N, G)), var.view((N, G)))
+        # TODO: test results!!!
+
+    #
+    # This test implements group normalization as 2 TC kernels using the
+    # tc.define_with_autograd abstraction. In particular it is ** NOT **
+    # possible to insert reshapes between TCs moments operate on a 5-D tensor
+    # (Note: loop collapsing modulo conforming strides is an easy TC-level
+    #  transformation with nice expected benefits).
+    # The tc.define_with_autograd API is more bloated than hoped but it is
+    # unclear we can do better, suggestions most welcome.
+    # Maybe consider making cache_filename and tuner_config global parameters.
+    #
+    def test_group_norm_2kernels_autograd(self):
+        group_normalization = """
+            def moments(float(N, G, D, H, W) I) -> (mean, var) {
+                # var = E(x^2) - mean^2.
+                mean(n, g) +=! I(n, g, r_d, r_h, r_w)
+                 var(n, g) +=! I(n, g, r_d, r_h, r_w) * I(n, g, r_d, r_h, r_w)
+                mean(n, g)  = mean(n, g) / (D * H * W)
+                 var(n, g)  =  var(n, g) / (D * H * W) - mean(n, g) * mean(n, g)
+            }
+
+            def group_normalization(
+                float(N, G, D, H, W) I, float(G, D) gamma, float(G, D) beta,
+                float(N, G) mean, float(N, G) var) -> (O)
+            {
+                O(n, g, d, h, w) = gamma(g, d)
+                    * ( I(n, g, d, h, w) - mean(n, g) )
+                    * rsqrt( var(n, g) + 1e-5 )
+                    + beta(g, d)
+            }
+        """
+
+        N, G, D, H, W = 32, 32, 4, 56, 56
+        cache_filename = "/tmp/" + str(uuid.uuid4())
+        group_norm = tc.define_with_autograd(
+            group_normalization,
+            forward_entry_points=("moments", "group_normalization"),
+            backward_entry_points=(),
+            # First: I, gamma, beta -> I
+            #        => (0, )
+            # Second: I, gamma, beta, mean, var -> I, gamma, beta, mean, var
+            #        => (0, 1, 2, 3, 4) => None
+            forward_input_indices=((0, ), ),
+            cache_filename=cache_filename,
+            tuner_config=tuner_config)
+        I = Variable(torch.randn(N, G, D, H, W).cuda())
+        gamma = Variable(torch.randn(G, D).cuda())
+        beta = Variable(torch.randn(G, D).cuda())
+        # First occurrence on a new cache filename triggers tuning
+        mean, var, out = group_norm(I, gamma, beta)
+        # Subsequent occurrences do not
+        mean, var, out = group_norm(I, gamma, beta)
+        # TODO: test results!!!
```

Comment:
It would be good to test that the tuning did actually happen somehow. We've got bitten by that when we were writing the JIT tests (after some changes, parts that were supposed to run some compiled versions weren't doing that)